Added e2e tests for alerting

skrech opened the pull request:
# Description

TBD

vicenteqa commented on test/e2e/cypress/pageObject/settings_po.js:
```diff
@@ -152,6 +155,15 @@ const _clickRemovePasswordButton = () =>
 const _clickRemoveSumaCaCertButton = () =>
   cy.get(sumaSettingsModal.removeCaCertButton).click();
 
+export const getAlertingEnabled = () => cy.get('[aria-label="alerting-enabled"]');
```
hey Kristiyan, this approach of how to get the elements doesn't match the one followed in the rest of the test suite. I considered doing it this way but with this we would miss the possibility of combining selectors or defining them dynamically when needed, basically having the plain strings in const, gives more freedom at the time of defining the test methods/assertions.

skrech commented on test/e2e/cypress/e2e/settings.cy.js:
```diff
@@ -170,6 +170,48 @@ context('Settings page', () => {
+      });
+
+      it('should be displayed correctly', () => {
+        settingsPage.getAlertingEnabled().should('have.text', 'Enabled');
```
Hey @vicenteqa, I actually saw that the rest of the test suite is structured differently. Anyway, I decided to go ahead and implement it a bit differently just to trigger a discussion. I'm ready to re-write it in the same way as the rest of the tests but just wanted to see what you guys think about slightly different perspective:

I'm actually a big proponent of the concept of "page objects" -- meaning, writing code that would abstract the tasks commonly needed when testing. However, putting the assertions in the page object is going too far for my taste :) I'll try to summarize why I think so in the following paragraphs.

The main problem for me is the "shallow" abstraction -- having name for every operation is more distracting than helping me out on understanding the code.   Eg. with `settingsPage.alertingUsernameHasExpectedValue() ` what is actually the expected value? I feel that sometimes the name of the function might be longer than actually writing the assertions verbatim. And more, I've randomly sampled several validation functions from the page object -- they were used only on one place. For me, following the simplicity principle, eg. extract into function only when the logic is duplicated, is better plan. 

Another small thing that I've noticed is that the name of the test is the same as the function it calls most of the time with small tests. In that case... it seems that we are reluctant to put any code in the tests, everything  should be in the page object. But why? 

I suspect that the reason for this seem to be to make the test cases read like English sentences. But that doesn't bring much to the table in my opinion:

- We have a small test suite file. But have an enormous page file.
- Maybe we protected our test suite from change of framework down the road by wrapping everything in a page object function or variable. But then, what is the guarantee that  the structure of the new framework won't change (eg. usage of `describe`, `it`, etc), or expect some context object that should be passed down, etc, etc.  

It seems to me that we've only increased the indirection without immediate benefit. Having abstraction for everything leads to extremely verbose code which is very time consuming to write :/ To me Cypress syntax is very expressive and find it concise as well. The `get(...).should(...)` syntax takes all the ambiguity out of the way. The places with `...` could be replaced with context-dependent variables -- either from page object getter function or constant or even a test suite constant (as opposed to page object one). Despite usage of page objects for encapsulating selectors and getters seems fine, I think that the pattern is envisioned and works best for user actions (typing into fields, series of actions taken, etc) and maybe we should concentrate on using it predominantly for that.

This was a lot of text, I'm sorry :) Anyway, my point is to raise the question of maybe relaxing the page object usage to at least not include the assertions. In my humble opinion, if there is one thing that shouldn't be missing from test case code it is the assertions. I don't find test suite files being long to be a problem, but definitely find big page object to resemble the famous _"god object"_ anti-pattern. That being said, I totally understand the style consistency concerns that might be pointed out. In that case, I'll re-write the code as the rest of the test suite. 

Please, let me know what you think about these remarks and sorry if at places I may sound harsh, I was trying to keep the text as concise as possible... but now as I look at the wall of text I see I might have failed at this :D

nelsonkopliku commented on guides/alerting/alerting.md:
````diff
@@ -107,6 +107,23 @@ testing during development.
+  smtp_username: nil,
+  smtp_password: nil
+```
+
````
Right, it's already there https://github.com/trento-project/web/pull/3461/files#diff-4f36118d05999cec87dec4895c8202881ba0577dc80fba4050aa50bbe9e326b3R96

gagandeepb commented on test/e2e/cypress/pageObject/settings_po.js:
```diff
@@ -694,9 +978,37 @@ export const updateApiKeyExpiration = (apiKeyExpiration) => {
   });
 };
 
+export const resetAlertingSettingsDB = () => {
+  cy.task('sqlExecute', {
+    query: "DELETE FROM settings WHERE type = 'alerting_settings'",
+  });
+};
```
@nelsonkopliku Sure, perhaps there's no difference in what ends they achieve (delete some entries from the db). But by that rationale, are you suggesting it is fine to replicate the data model of the database in the UI focused E2E suite as a general thing?

skrech replied:
There is definitely value in hiding implementation details, or at least not exposing them to entities that don't necessarily need to know about them. However, this comes with it's own set of trade-offs -- I could imagine that cypress runner and trento-web might be running on different machines/containers, which would impede usage of `trento eval` (if I understand correcly, this is the way to execute test utils in the BE?), I think. 
I took the easiest way around this, directly accessing the DB. It's not perfect but I think it's good trade-off for something so small as 4-5 alerting tests.

nelsonkopliku replied:
@gagandeepb No, I am not suggesting to replicate the data model 😄 I agree on your general sentiment.

What I am saying is that what proposed looks a reasonable tradeoff considering the constraints (not going to reiterate on those)

The immediate alternatives to avoid the explicit query do not sound much better either
- resetting the whole db to start from a clean sheet would entail impact on test performance
- implement the clean feature for the sake of testing would mean what @skrech outlined already

We're not in a backend embedded e2e test suite, and I don't see such a move to happen soonish.

A middle ground could be having cypress avoid running the raw query by executing a mix task utility, which internally would call a Repo.delete, not much more. And I'd be open to consider that option.

skrech commented:
Added the CI changes. Instead of creating a new Job, though, I went for re-using the `test-integration` job. It's definition is generic enough and the only redundant step is starting the `idp` profile for docker compose (but this causes no harm for alerting). 
I'd even suggest we rename the `test-integration` job to something more generic like `e2e-parameterized`, for example - effectively it takes all the steps necessary to parameterize e2e test -- setting cypress env-vars, modifying the compile-time elixir configuration and specifying what spec files to run. We can also add the docker compose profile in the matrix and then it would become truly universal. 
I had to re-arrange  `settings.cy.js` for better skip handling. It turned out in the case where we're using the test-integration job the subset of desired alerting tests are run along with the tests for suma, activity log, etc. To avoid that with as little change to existing files as possible, I decided to split the alerting tests involving DB interaction into their own file `settings_integration.cy.js` following the de facto convention established by `sso_integration.cy.js`. This works great, IMO apart from the somewhat misleading name -- `settings_integration`, there is no integration actually going on. 
I also added a small protection mechanism to skip the enforced-from-env test if they are run against local development BE target currently using the DB for configuration.
With that, this PR seems to be complete. All CI checks are passing. If there is a consensus on leaving the SQL excerpt as is, then I think it's in a good shape to be merged.

nelsonkopliku commented on mix.exs:
```diff
@@ -242,6 +242,6 @@ defmodule Trento.MixProject do
 
   defp get_version_from_file do
-    File.cwd!() |> Path.join("VERSION") |> File.read!() |> String.trim()
+    __DIR__ |> Path.join("VERSION") |> File.read!() |> String.trim()
```
What issues was the previous function usage causing? 
Not that I am against the change, just to understand 😄

skrech replied:
Very good catch :) I even intent to leave this into it's own commit when merging this.

It can't be seen from the diff of the PR but I tried very hard to actually start the Mix task from different current working directory. And when specifying custom path to `mix.exs`, reading the files wont' succeed because it won't find VERSION file in the what happens to be the current directory. `__DIR__` would work because it's bound to the location of the file being read, so it's very robust.

nelsonkopliku reviewed (approved):
Nice, thanks!
